pygsti/tools/optools.py

```diff
@@ -81,41 +68,92 @@ def fidelity(a, b):
     float
         The resulting fidelity.
     """
-    evals, U = _np.linalg.eig(a)
-    if len([ev for ev in evals if abs(ev) > 1e-8]) == 1:
+    __SCALAR_TOL__ = _np.finfo(a.dtype).eps ** 0.75
+    # ^ use for checks that have no dimensional dependence; about 1e-12 for double precision.
+    __VECTOR_TOL__ = (a.shape[0] ** 0.5) * __SCALAR_TOL__
+    # ^ use for checks that do have dimensional dependence (will naturally increase for larger matrices)
+    def assert_hermitian(mat):
+        hermiticity_error = _np.abs(mat - mat.T.conj())
+        if _np.any(hermiticity_error > __SCALAR_TOL__):
+            message = f"""
+                Input matrix 'mat' is not Hermitian, up to tolerance {__SCALAR_TOL__}.
+                The absolute values of entries in (mat - mat^H) are \n{hermiticity_error}. 
+            """
+            raise ValueError(message)
+    
+    assert_hermitian(a)
+    assert_hermitian(b)
+
+    def check_rank_one_density(mat):
```

Comment:
Could you add a sentence or two explaining the logic to the comments for this function (or a reference to the algorithm if it is well known). I'm going to try and parrot back what I think the logic is here, along with some stuff I am confused by, to see if I understood correctly (I'm going to do so with bra-ket notation, so I apologize). Assume mat is a rank 1 matrix. If so we can write it as a rank 1 projector |v><v|. Let |w> be a random test vector. Applying mat to |w> we get  <v|w>*|v>, which can can renormalize to get what you called candidate_v. I.e. if our assumption is correct then candidate_v = |v>. alpha is the (candidate_v, ) matrix element of mat: <cand_v|mat|cand_v>.  Finally, take the outer product of |cand_v> with itself and compare to mat. If these are sufficiently close it must've been the case that mat was indeed a rank 1 projector, otherwise we have a proof by contradiction that mat was not a rank 1 projector. 

Questions: 
1. Is there any reason why the test vector is chosen randomly? I would have thought that any test vector with full support would have worked, e.g. the all-ones vector.
2. Why do we need alpha here? If I described this correctly then shouldn't alpha always be 1 in the case where mat is a rank 1 projector?

Reply:
@coreyostrove I've updated this code to be more readable and to include comments. Take another look?

Reply:
Thanks, I had thought this looked a lot like you were doing a power-iterationesque algorithm here. I'm admittedly still not sure why randomness is necessary here, as all for the power iteration all you need is non-trivial overlap with the dominant eigenspace. But it is clear to me this is sufficient, if not necessary, so I'm cool with suspending disbelief. Thanks for adding the additional explanation (future us will thank you!).